rmrk-token/src/multiresource.rs

```diff
@@ -0,0 +1,248 @@
+    /// * `metadata_uri`:  a string pointing to a metadata file associated with the resource
+    pub async fn add_resource_entry(
+        &mut self,
+        id: u8,
+        src: String,
+        thumb: String,
+        metadata_uri: String,
+    ) {
+        // sends message to resource storage contract
+        add_resource_entry(&self.resource_id, id, src, thumb, metadata_uri).await;
+        msg::reply(RMRKEvent::ResourceEntryAdded { id }, 0).unwrap();
+    }
+
+    /// Adds resource to an existing token
+    /// Proposed resource is placed in the "Pending" array
+    /// A pending resource can be also proposed to overwrite an existing resource
+    ///
+    /// Requirements
+    /// Token with indicated `token_id` must exist
+    /// The proposed resource must not already exist for the token
+    /// The resource that is proposed to be overwritten must exist for the token
+    ///
+    /// Arguments:
+    /// * `token_id`: an id of the token
+    /// * `resource_id`: a proposed resource
+    /// * `overwrite_id`: a resource to be overwritten
+    pub async fn add_resource(&mut self, token_id: TokenId, resource_id: u8, overwrite_id: u8) {
+        self.assert_token_does_not_exist(token_id);
+        assert_resource_exists(&self.resource_id, resource_id).await;
+
+        if let Some(token_resources) = self.multiresource.token_resources.get(&token_id) {
+            assert!(
+                !token_resources.contains(&resource_id),
+                "Resource already exists on token"
+            );
+        }
+        if let Some(pending_resources) = self.multiresource.pending_resources.get(&token_id) {
+            assert!(
+                pending_resources.len() < MAX_RESOURCE_LEN as usize,
+                "Max pending resources reached"
+            );
+        }
+
+        if overwrite_id != 0 {
+            if let Some(token_resources) = self.multiresource.active_resources.get(&token_id) {
+                assert!(
+                    token_resources.contains(&overwrite_id),
+                    "Proposed overwritten resource must exist on token"
+                );
+            } else {
+                panic!("No resources to overwrite")
+            }
+            self.multiresource
+                .resource_overwrites
+                .entry(token_id)
+                .and_modify(|r| {
+                    r.insert(resource_id, overwrite_id);
+                })
+                .or_insert_with(|| {
+                    let mut r = BTreeMap::new();
+                    r.insert(resource_id, overwrite_id);
+                    r
+                });
+        }
+
+        self.multiresource
+            .token_resources
+            .entry(token_id)
+            .and_modify(|r| {
+                r.insert(resource_id);
+            })
+            .or_insert_with(|| {
+                let mut r = BTreeSet::new();
+                r.insert(resource_id);
+                r
+            });
+        self.multiresource
```

Comment:
I think you can write one function for inserting one resource, in this function you do it many times